Multi source terrain provider

juunie-roh opened the pull request:
<!--
Thanks for the Pull Request!

Please review [Contribution Guide](https://github.com/CesiumGS/cesium/blob/main/CONTRIBUTING.md) before opening your first Pull Request.

To ensure your Pull Request is reviewed and accepted quickly, please refer to our [Pull Request Guidelines](https://github.com/CesiumGS/cesium/blob/main/CONTRIBUTING.md#pull-request-guidelines).

-->

# Description

<!-- Describe your changes in detail -->

This is an initial draft of migration from external library TypeScript implementation into Cesium development environment.

The naming conventions are not considered in this draft.

<!-- Consider: Why is this change required? What problem does it solve? -->

You can see the detailed description about this feature at the related issue, #12817.

<!-- Include screenshots if appropriate -->

## Issue number and link

<!-- If it fixes an open issue, link to the issue here -->

This PR relates to #12817.

<!-- Consider: If suggesting a new feature or change, discuss it in an issue first. -->

## Testing plan

<!-- Describe in detail how you tested your changes. If this fixes a bug, list the steps to reproduce the original issue. -->

You can see conducted unit tests at [external library repository](https://github.com/juunie-roh/cesium-utils/blob/main/src/terrain/hybrid-terrain-provider.test.ts).

# Author checklist

- [x] I have submitted a Contributor License Agreement
- [x] I have added my name to `CONTRIBUTORS.md`
- [ ] I have updated `CHANGES.md` with a short summary of my change
- [ ] I have added or updated unit tests to ensure consistent code coverage
- [ ] I have updated the inline documentation, and included code examples where relevant
- [x] I have performed a self-review of my code

juunie-roh committed:
chore: add name

juunie-roh committed:
feat: Initial translation of TypeScript library feature

juunie-roh committed:
docs: Update usage example

juunie-roh committed:
Merge branch 'main' into multi-source-terrain-provider

juunie-roh committed:
Merge branch 'main' into multi-source-terrain-provider

juunie-roh committed:
Merge branch 'CesiumGS:main' into multi-source-terrain-provider

juunie-roh committed:
Merge branch 'main' into multi-source-terrain-provider

juunie-roh committed:
Merge branch 'main' into multi-source-terrain-provider

juunie-roh commented:
Hi, @ggetz ! 👋 
You're right, that this is not for showing multiple terrains at once.
My specific use case was to **integrate terrains** from different sources **seamlessly**.
I set a global terrain fallback, then configured the specific area where needed to show terrain with different source.(higher resolution could be one of the example)
This implementation solves the problem by delegating `requestTileGeometry` to different terrain providers based on tile coordinates and zoom levels.

https://juun.vercel.app/cesium-utils/terrain
I have working demonstration here. The global fallback terrain with `EllipsoidTerrainProvider`, and configured area with Cesium's default provider at zoom level 13. (red-colored area).

juunie-roh committed:
Merge branch 'main' into multi-source-terrain-provider

ggetz commented:
Thanks @juunie-roh! Your demonstration looks to be working great.

We'll take a closer look at this PR and review shortly. Please let us know if there are any updates to the status of this PR in the meantime.

juunie-roh committed:
Merge branch 'main' into multi-source-terrain-provider

juunie-roh committed:
Merge branch 'main' into multi-source-terrain-provider

juunie-roh committed:
Merge branch 'main' into multi-source-terrain-provider

lukemckinstry commented:
> potential terrain cracking issues (skirt generation) at the boundaries of different sources.

There is logic in the terrain provider to handle skirt generation where adjacent tiles are at different LODs and cracks appear, so you could reference that. 

> resource management when handling multiple providers 

I am less certain what complications may arise, at a high level if you follow the control flow for the single terrain provider - destroy resources when they go out of use - I think it will be ok, but keep an eye out for lower level optimizations in the current terrain code.

juunie-roh committed:
Merge branch 'main' into multi-source-terrain-provider

juunie-roh commented:
The [demo link](https://juun.vercel.app/ko/cesium-utils/terrain) is back online.

<img width="1454" height="897" alt="image" src="https://github.com/user-attachments/assets/c6f7df3e-5f13-48ba-a244-993c2c93819c" />

From the thick yellow lines at the edges of terrain tiles from the image, I guess the skirt generation is working, and cracks appear just because the height difference is larger than the default `skirtHeight`. (Cesium world terrain vs Ellipsoid)

```js
const skirtHeight = provider.getLevelMaximumGeometricError(level) * 5.0
```

I'm also refining other properties (credit, errorEvent, hasVertexNormals, hasWaterMask) to ensure full compatibility. I’ll push the update as soon as it's ready!

juunie-roh committed:
HybridTerrainProvider: apply updated logics with enhanced property controls

juunie-roh committed:
HybridTerrainProvider: Add temporary demo executable locally with build output

juunie-roh committed:
Merge branch 'main' into multi-source-terrain-provider

juunie-roh commented:
Pushed with temporary demo identical to the external demo link, at root `HybridTerrainProviderDemo.html`

Few things:

- `credit` → concatenates all credits from registered providers (if present)
- `hasVertexNormals` & `hasWaterMask` → return true if any of registered provider has one.
- `errorEvent` → added listeners to each registered providers to forward all possible errors.

juunie-roh committed:
Merge branch 'main' into multi-source-terrain-provider

juunie-roh committed:
HybridTerrainProvider: Replace TypeScript tuple type annotation with plain `number[]`

juunie-roh committed:
HybridTerrainProvider: Refactor `HybridTerrainProvider.TerrainRegion.contains` to `HybridTerrainProvider.regionContainsTile`

Renamed `contains` static utility to avoid "Duplicate identifier" error while running `build-ts` script.

juunie-roh committed:
HybridTerrainProvider: Format temporary demo to match sandcastle

juunie-roh committed:
Merge branch 'main' into multi-source-terrain-provider

juunie-roh committed:
HybridTerrainProvider: Apply updated method name

juunie-roh committed:
HybridTerrainProvider: Remove `fallbackProvider`

juunie-roh committed:
HybridTerrainProvider: Remove stale comments

juunie-roh commented:
I have formatted the demo under "multiple-terrain-sources", and updated jsDocs to pass scripts `build-docs` and `build-ts`.

- The terrain source I used in production is proprietary, so I'm afraid I cannot use it for the public demonstration. Do you have a sample detailed terrain asset from Cesium ion I could use for the demo that can comparable with CesiumWorldTerrain?
- `this._fallbackProvider` was intended to be a safety net for an asnyc-ready model failure, but since the provider now only reaches ready when all sub-providers are ready, that intent no longer applies. I have simplified those into two options as you suggested.

**On the clipping-planes workflow comparison**:

I considered a `Rectangle` based API during development and withdrew it for several reasons.

1. **As far as I know**, there is no runtime mechanism to synthesize partial tile data. 
2. The `TerrainProvider`'s contract assumes complete terrain tiles as the atomic unit.
3. `TerrainProvider` doesn't expose raw elevation data during runtime.

3D Tiles, on the other hand, are possible to replace clipped area because it handles geometric information during runtime directly.

In my case, I was able to control the terrain source by lon/lat level. I used [Cesium Inspector](https://sandcastle.cesium.com/?id=cesium-inspector) — terrain section — to track down tile coordinates and its lon/lat ranges.

(p.s. The provider assumes each sub-providers to have the same `TilingScheme`)
